### PR TITLE
Sct initial push

### PR DESCRIPTION
**Pull request by syllamoh**
RSR-31 / Create a new SCD file with Header
EPIC RSR-51 / RSR-6 : S79 : Add DataTypeTemplates from ICD to current SCD
EPIC RSR-51 / RSR-60 : Add a new IED from ICD to current SCD
EPIC RSR-51 / RSR-64 : Associate a list of Connected Access Point (ConnectedAP) to a SubNetwork
RSR-86 COMPAS List for a given signal ExtRef, all possible IED, LDevice, LN and DO that can be defined as source
RSR-81 COMPAS List all External References signals ExtRef belonging to an IED and LDevice.inst of an SCD file
EPIC RSR-51 / RSR-62 : create a new SubNetwork to current SCD

### CODE REVIEW

**Comment by Flurb** [comment]
How do I correctly review this?
It's a pretty big PR!

**Comment by syllamoh** [comment]
Hi
Who is merging commit on branch under pull request ?

**Comment by syllamoh** [comment]
Plus when doing stuff like this (which is totally not recommended), make sure you do test.

**Comment by Flurb** [comment]
Hi @syllamoh we saw that the SonarCloud action wasn't added yet. So we added it :)
Otherwise we are merging code into the develop branch that hasn't been scanned. Not something we want!

**Comment by Flurb** [comment]
I'm also going to add a build Github Action (like the one in compas-cim-mapping repository https://github.com/com-pas/compas-cim-mapping/actions/workflows/compas-ci.yml) to check if the projects build correctly.
Right now, it doesn't. We had to check it manually, this is not something we want.

**Comment by syllamoh** [comment]
Hi @Flurb 
That's not how collaborative development work.
First, if you want to add code or document on existing branch, you create issue and do pull request.
Second, you don't add change to branch under pull request whom you're not the requester
Third the least you can do in general, is to test your change before merging. If you did this you'd have notice that your script buggy.

Worst this pull request target is the branch develop, and you merge a code from develop to this branch. I'm sorry, but this is not OK at all.
It's better you remove what you added, and follow the process correctly.

**Comment by syllamoh** [comment]
@Sander3003 
No problem for me.
It was just very surprising the way things were done.

**Review by dlabordus — CHANGES_REQUESTED** [review]
- sct-data: Some of the DTO Classes in this project aren't used, only by test-classes.
- sct-data: H2XMLTypeMapper is only used in the test part, why not put it in "src/test/java"
- sct-data: There are Test XML Files in "src/test/resource", but I don't think there are used in this project.
- sct-data: I think the idea is that ScdJpaRepository also implements IScdCrudRepository, and that this interface is used in de Service Layer. Now ScdJpaRepository is used in the Service Layer.
- sct-service: SimpleSqlScdService is working with the class SimpleSCD, but I see SimpleSCD as a data object how stuff is stored. Should we build a Service that will be handling the SCL and how this is stored is up to a Storage Service. In that Storage Service class we can use SimpleSCD.
- sct-service: Also, SimpleSqlScdService has a lot of extra methods add*, extract*. These are only working on the SCL class, so by passing the SimpleSCD a lot of extra Unmarshalling and Marshalling is done. It looks like these methods are used by some GUI Interface, but that's unclear to me. And it's now becoming quite a big service with a lot of responsibilities. We prefer smaller services.
- sct-service: In the directory "src/main/resources" there are xml files, that probably should be there. I think there are by mistake copied there because they are also in "src/test/resources".
- sct-service: currently this project only contains a Singletone Spring Bean, and it can't be access as a microservice through a REST Interface.

**Comment by dlabordus** [comment]
> @dlabordus
> 
> > * sct-data: Some of the DTO Classes in this project aren't used, only by test-classes.
> 
> This not the final version for compas-sct. Developement is ongoing. This pull request only addressed issue mentioned before. nothing else. DTO classes as you see are model(sct-data), will be used in sct-data and in the future sct-app (coming soon).
> 
I would prefer to first have the basic in place and then add these classes when needed, then it's also clear how these should look.
> > * sct-data: I think the idea is that ScdJpaRepository also implements IScdCrudRepository, and that this interface is used in de Service Layer. Now ScdJpaRepository is used in the Service Layer.
> 
> Almost, but not really. I'm writing the conceptual architecture. ScdJpaRepository is a spring JPA interface. There ScdRepository is class that implements IScdCrudRepository, that class delegate DAO operation to spring. Any correct IDE can help you navigate.
Oke, I will wait for that and see how we can integrate the BaseX version, because that's not JPA based.
> 
> > * sct-service: SimpleSqlScdService is working with the class SimpleSCD, but I see SimpleSCD as a data object how stuff is stored. Should we build a Service that will be handling the SCL and how this is stored is up to a Storage Service. In that Storage Service class we can use SimpleSCD.
> 
> SimpleSCD has no storage capability, juste simple POJO (the SCL and meta data)
> SimpleSqlScdService implements IScdService. It contains the DAO class that deal with storage. I suppose you're familiar with generic programing.
SimpleSCD is a Pojo, but contains all kind of JPA Annotations, but in BaseX these aren't usefull.
> 
> > * sct-service: Also, SimpleSqlScdService has a lot of extra methods add*, extract*. These are only working on the SCL class, so by passing the SimpleSCD a lot of extra Unmarshalling and Marshalling is done. It looks like these methods are used by some GUI Interface, but that's unclear to me. And it's now becoming quite a big service with a lot of responsibilities. We prefer smaller services.
> 
> SimpleSqlScdService has oe single mission : manipulate SCD file. It has no direct interaction with any GUI.
> As the XML manipulation is managed by JAXB facilities, one should indeed marshal raw xml data fromDB, process it, the unmarshal to put it back to DB. They are done once par service call.
Maybe it's nice to have a separate Service to manipulate the SCL directly. And only marchalling is then needed when saving to a Database.

**Comment by Flurb** [comment]
core-commons dependency needs to be added! :)

**Comment by Sander3003** [comment]
#6

File: sct-commons/pom.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    <version>local-SNAPSHOT</version>
+
+    <properties>
+        <sonar.coverage.jacoco.xmlReportPaths>${basedir}/${aggregate.report.dir}</sonar.coverage.jacoco.xmlReportPaths>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-context</artifactId>
```
Comment (Flurb) [inline sct-commons/pom.xml:27]: Dependency not being used anymore?

File: sct-commons/pom.xml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.sun.xml.bind</groupId>
+            <artifactId>jaxb-impl</artifactId>
+            <version>2.3.3</version>
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-oxm</artifactId>
```
Comment (Flurb) [inline sct-commons/pom.xml:43]: Dependency not being used anymore?

**Review by Flurb — CHANGES_REQUESTED** [review]

**Review by Flurb — COMMENTED** [review]
Also, the files in sct-commons/src/test/resources/ are all empty, why are they added?

File: sct-commons/pom.xml (hunk 1 above)
Comment (syllamoh) [reply sct-commons/pom.xml:27]: Do you guys have a white liste of libraries that RTE should use or not use.

There's also the following package in the POM, that are not used in this pull request. It will be certainly used in the next:
 
<dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
            <version>1.18.16</version>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-api</artifactId>
            <version>5.7.0</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.hamcrest</groupId>
            <artifactId>hamcrest</artifactId>
            <version>2.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.mockito</groupId>
            <artifactId>mockito-core</artifactId>
            <version>3.6.28</version>
            <scope>test</scope>
        </dependency>
        
Even the only class defined is useless at this moment. 

What's the criteria for a pull request.
Do we need weeks of discussion because a non releasable pull request contains empty conf file ?

Comment (Flurb) [reply sct-commons/pom.xml:27]: It's useless to add dependencies which are not being used.
You can better delete them :)
If they will be used in the next pull request, you can add them at that point.

By leaving dependencies in, you can forget about them and they will stay in even if they are not being used.

My advice would be: Delete them now, add them later if needed.

Comment (Flurb) [reply sct-commons/pom.xml:27]: Same with the config files by the way!
I see 2 empty files are added. They are useless right now, delete them.

If it's needed in the future, you can add them at that time.

Right now, it's just a waste.

Comment (syllamoh) [reply sct-commons/pom.xml:27]: I bear wasting my time ... that's not a big deal for me at this moment

THe aim is to be sure all imports work with the build script (internal dependencies, external dependencies etc ... ). I guess that's understandable.

Comment (Flurb) [reply sct-commons/pom.xml:27]: Allright, it will be fixed in the future then! :)

**Review by Flurb — APPROVED** [review]

**Comment by Flurb** [comment]
You can rebase/merge it @syllamoh
